libdnf5/base/goal.cpp

```diff
@@ -2206,6 +2412,138 @@ GoalProblem Goal::Impl::add_reason_change_to_goal(
     return GoalProblem::NO_PROBLEM;
 }
 
+GoalProblem Goal::Impl::resolve_reverted_transactions(base::Transaction & transaction) {
+    if (!revert_transactions) {
+        return GoalProblem::NO_PROBLEM;
+    }
+    auto ret = GoalProblem::NO_PROBLEM;
+
+    using Action = transaction::TransactionItemAction;
+    using Reason = transaction::TransactionItemReason;
+    const std::unordered_map<Action, Action> REVERT_ACTION = {
+        {Action::INSTALL, Action::REMOVE},
+        {Action::UPGRADE, Action::REPLACED},
+        {Action::DOWNGRADE, Action::REPLACED},
+        {Action::REINSTALL, Action::REINSTALL},
+        {Action::REMOVE, Action::INSTALL},
+        {Action::REPLACED, Action::INSTALL},
+        {Action::REASON_CHANGE, Action::REASON_CHANGE},
+    };
+    transaction::TransactionReplay replay;
+    auto history = base->get_transaction_history();
+
+    auto & [reverting_transactions, settings] = *revert_transactions;
+    //TODO(amatej): Implement merging of transactions and merge the vector
+    //              instead of taking the first one.
+    auto & reverting_transaction = reverting_transactions[0];
+
+    for (const auto & pkg : reverting_transaction.get_packages()) {
+        transaction::PackageReplay package_replay;
+        package_replay.nevra = libdnf5::rpm::to_nevra_string(pkg);
+        auto reverted_action = REVERT_ACTION.find(pkg.get_action());
+        libdnf_assert(
+            reverted_action != REVERT_ACTION.end(),
+            "Cannot revert action: \"{}\"",
+            transaction_item_action_to_string(pkg.get_action()));
+        package_replay.action = reverted_action->second;
+
+        // We cannot tell the previous reason if the action is REASON_CHANGE it could have been anything.
+        // For reverted action INSTALL and reason CLEAN the previous reason could have been either DEPENDENCY or WEAK DEPENDENCY
+        // to pick the right one we have to look into history.
+        if ((package_replay.action == Action::REASON_CHANGE) ||
+            (package_replay.action == Action::INSTALL && pkg.get_reason() == Reason::CLEAN)) {
+            auto nevras = rpm::Nevra::parse(pkg.to_string(), {rpm::Nevra::Form::NEVRA});
```

Comment:
I believe that `nevras` object is not required for following reasons:

1. It was already created above `package_replay.nevra = libdnf5::rpm::to_nevra_string(pkg);`
2. It is not required, because it can be used already parsed form
` package_replay.reason = history->transaction_item_reason_at(pkg.get_name(), pkg.get_arch(), reverting_transaction.get_id() - 1);`

Reply:
Or i there any other reason why you want to construct and parse NEVRA from transaction pkg object?

Reply:
There is no valid reason here I will remove it. Thank you.